src/main.rs

```diff
@@ -539,30 +545,53 @@ fn notmain() -> Result<i32, anyhow::Error> {
 
     let debug_frame = debug_frame.ok_or_else(|| anyhow!("`.debug_frame` section not found"))?;
 
-    // print backtrace
-    let top_exception = backtrace(
-        &mut core,
-        pc,
-        debug_frame,
-        &elf,
-        &vector_table,
-        &sp_ram_region,
-        &live_functions,
-        &current_dir,
-    )?;
+    print_separator();
+
+    // check if the user enforced a backtrace print
+    let force_backtrace = match opts.force_backtrace.as_deref() {
+        Some("False") | Some("0") | Some("false") | None => false,
+        _ => true
+    };
+
+    let top_exception = construct_backtrace(
+            &mut core,
+            pc,
+            debug_frame,
+            &elf,
+            &vector_table,
+            &sp_ram_region,
+            &live_functions,
+            &current_dir,
+            // TODO any other cases in which we should force a backtrace?
+            force_backtrace || canary_touched
+        )?;
 
     core.reset_and_halt(TIMEOUT)?;
 
     Ok(
-        if let Some(TopException::HardFault { stack_overflow }) = top_exception {
-            if stack_overflow {
+        match top_exception {
+            Some(TopException::StackOverflow) => {
                 log::error!("the program has overflowed its stack");
+                SIGABRT
             }
+            Some(TopException::HardFault) => {
+                log::error!("the program panicked");
+                SIGABRT
+            }
+            None => {
+                if !force_backtrace {
+                    let mut re_run_args = "PROBE_RUN_BACKTRACE=1 probe-run".to_string();
+                    for arg in std::env::args().skip(1) {
+                        re_run_args = format!("{} {}", re_run_args, arg);
+                    }
```

Comment:
This way of constructing the string seems a bit hacky. Is there maybe a more elegant iterator-based one?

Reply:
As mentioned later we can maybe use `String::push_str`?